### PR TITLE
refactor!: Change PodBuilder::termination_grace_period to take Duration struct

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Part of https://github.com/stackabletech/issues/issues/357

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: CHANGELOG.md
Comment (nightkr) [inline CHANGELOG.md:10]:
```suggestion
- BREAKING: Rename `PodBuilder::termination_grace_period_seconds` to `termination_grace_period`, and change it to take `Duration` struct ([#672]).
```

**Review by Techassi — COMMENTED** [review]
Looks good in general. I added a small comment about the upstream change. This can be approved when the upstream decision is made.

File: src/error.rs
Comment (nightkr) [inline src/error.rs:116]: Is too long for what? Where does the duration come from? What is the limit? Where's the source error?

Comment (nightkr) [reply src/error.rs:116]: If we expect the downstream operator to provide this context, make a separate error type instead of adding even more crap to the overbloated corpse that is `OperatorError`.

Comment (sbernauer) [reply src/error.rs]: Please have another look.

File: src/builder/pod/mod.rs
Comment (sbernauer) [reply src/builder/pod/mod.rs]: I fear this is a bigger thing. We have `Option<i32>` or `Option<i64>` multiple time in k8s-openapi.
E.g. the number of STS replicas is an i32 as well: https://github.com/stackabletech/hdfs-operator/blob/9830327fc493bca5f7d556ba34e04e6fa15ab587/rust/operator/src/hdfs_controller.rs#L699

Comment (nightkr) [reply src/builder/pod/mod.rs]: Yeah it'd be nice to resolve eventually but I wouldn't block on it.

Comment (Techassi) [reply src/builder/pod/mod.rs]: Yes. This is clearly a bigger issue, but fundamental issues lies within the OpenAPI spec, which doesn't define any *unsigned* integers, only signed ones (with different lengths). This can be fixed at different links in the chain:

- By K8s itself, by either specifying a `min` value of `0` or unsigned integers, by introducing a custom format or by using a JSON schema
- By k8s-openapi, which would manually fix known fields by using `u32` or `u64`. One example would be fields ending with `Seconds`.

File: src/builder/pod/mod.rs
Comment (nightkr) [inline src/builder/pod/mod.rs:33]:
```suggestion
    TerminationGracePeriodTooLong {
```

The error message notes the field specifically, so we should do the same in code to avoid confusion.

File: src/builder/pod/mod.rs
Comment (nightkr) [inline src/builder/pod/mod.rs:32]:
```suggestion
    #[error("termination grace period is too long (got {duration}, maximum allowed is {max})", max = Duration::from_secs(i64::MAX as u64))]
```

File: src/builder/pod/mod.rs
Comment (nightkr) [inline src/builder/pod/mod.rs:30]:
```suggestion
#[derive(Debug, thiserror::Error)]
```

Error types are typically not `PartialEq`.

Comment (sbernauer) [reply src/builder/pod/mod.rs:30]: Yeah, that's a leftover of my assert_eq!() test, which didn't work because of `TryFromIntError`

**Review by nightkr — APPROVED** [review]

File: src/builder/pod/mod.rs
Comment (Techassi) [reply src/builder/pod/mod.rs]: I added this issue to our backlog. We will tackle this in the future.

**Review by Techassi — APPROVED** [review]

**Comment by sbernauer** [comment]
Thanks for the review!

**Pull request merged**
